File: Q4_0/Denoise_test.m
%% ============================================================
% 降噪效果检查脚本：频谱对比 + 信噪比输出 (含包络包裹填色)
% 修正版：修复 MinPeakDistance 错误
% ============================================================
clear; clc; close all;

%% ---- 1. 读取数据 ----
filename = 'Filtered_Result.csv';
data = readtable(filename);
raw = data.RawDisplacement;
filtered = data.FilteredDisplacement;

valid = ~isnan(raw) & ~isnan(filtered);
raw = raw(valid);
filtered = filtered(valid);
N = length(raw);
fprintf('有效数据点数：%d\n', N);

%% ---- 2. 计算信噪比 ----
residual = raw - filtered;
SNR = 20 * log10(std(raw) / std(residual));
fprintf('信噪比 SNR = %.2f dB\n', SNR);

%% ---- 3. 功率谱估计 ----
window = min(256, N);
noverlap = round(window/2);
nfft = 1024;

[Pxx_raw, f] = pwelch(raw - mean(raw), hamming(window), noverlap, nfft, 1);
[Pxx_filt, ~] = pwelch(filtered - mean(filtered), hamming(window), noverlap, nfft, 1);

%% ---- 4. 计算上下包络线（修正后的函数） ----
[f_upper_raw, P_upper_raw] = getEnvelope(f, Pxx_raw, 'upper');
[f_lower_raw, P_lower_raw] = getEnvelope(f, Pxx_raw, 'lower');

[f_upper_filt, P_upper_filt] = getEnvelope(f, Pxx_filt, 'upper');
[f_lower_filt, P_lower_filt] = getEnvelope(f, Pxx_filt, 'lower');

%% ---- 5. 绘图 ----
figure('Position', [100, 100, 1000, 650]);
ax = gca;
set(ax, 'YScale', 'log');
hold on;

% 降噪后填充
fill([f_upper_filt; flipud(f_lower_filt)], ...
     [P_upper_filt; flipud(P_lower_filt)], ...
     'r', 'FaceAlpha', 0.08, 'EdgeColor', 'none');
plot(f_upper_filt, P_upper_filt, 'r--', 'LineWidth', 1.0);
plot(f_lower_filt, P_lower_filt, 'r--', 'LineWidth', 1.0);
plot(f, Pxx_filt, 'r-', 'LineWidth', 1.2);

% 降噪前填充
fill([f_upper_raw; flipud(f_lower_raw)], ...
     [P_upper_raw; flipud(P_lower_raw)], ...
     'b', 'FaceAlpha', 0.08, 'EdgeColor', 'none');
plot(f_upper_raw, P_upper_raw, 'b--', 'LineWidth', 1.0);
plot(f_lower_raw, P_lower_raw, 'b--', 'LineWidth', 1.0);
plot(f, Pxx_raw, 'b-', 'LineWidth', 1.2);


xlabel('频率 (周期/序号)');
ylabel('功率谱密度');
title(sprintf('降噪前后傅里叶频谱对比 (信噪比 = %.2f dB)', SNR));
legend('降噪后包络填充','降噪后上包络','降噪后下包络','降噪后功率谱',...
        '降噪前包络填充','降噪前上包络','降噪前下包络','降噪前功率谱',...
       'Location','best');
grid on;

%% ============================================================
% 辅助函数：提取频谱的上下包络（修正版）
% ============================================================
function [f_env, env_vals] = getEnvelope(f, Pxx, mode)
    % 按样本索引找峰值，避免 MinPeakDistance 单位问题
    minPeakDistance = max(3, round(length(f)/50));  % 样本点间隔
    if strcmp(mode, 'upper')
        [pks, idx] = findpeaks(Pxx, 'MinPeakDistance', minPeakDistance);
    else
        [pks, idx] = findpeaks(-Pxx, 'MinPeakDistance', minPeakDistance);
        pks = -pks;
    end
    locs = f(idx);
    % 插值到原频率网格
    if length(locs) >= 2
        env_vals = interp1(locs, pks, f, 'pchip', 'extrap');
    else
        env_vals = ones(size(f)) * mean(Pxx);
    end
    f_env = f;
end

File: Q2/denoise_test.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import welch, find_peaks
from scipy.interpolate import CubicSpline
import os

# ===== 设置 =====
plt.rcParams['font.sans-serif'] = ['Microsoft YaHei', 'SimHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

script_dir = os.path.dirname(os.path.abspath(__file__))
output_dir = os.path.join(script_dir, "Fourier")
os.makedirs(output_dir, exist_ok=True)

# ===== 读取数据 =====
df = pd.read_csv(os.path.join(script_dir, 'Filtered_Result.csv'))
raw = df['RawDisplacement'].values.astype(float)
filtered = df['FilteredDisplacement'].values.astype(float)

valid = ~np.isnan(raw) & ~np.isnan(filtered)
raw = raw[valid]
filtered = filtered[valid]
N = len(raw)
print(f'有效数据点数：{N}')

# ===== 信噪比 =====
residual = raw - filtered
SNR = 20 * np.log10(np.std(raw) / np.std(residual))
print(f'信噪比 SNR = {SNR:.2f} dB')

# ===== 功率谱 =====
window = min(256, N)
noverlap = window // 2
nfft = 1024

f, Pxx_raw = welch(raw - np.mean(raw), fs=1.0, window='hamming',
                   nperseg=window, noverlap=noverlap, nfft=nfft)
_, Pxx_filt = welch(filtered - np.mean(filtered), fs=1.0, window='hamming',
                    nperseg=window, noverlap=noverlap, nfft=nfft)

# ===== 包络函数 =====
from scipy.interpolate import PchipInterpolator

def get_envelope(f, Pxx, mode='upper'):
    min_dist = max(3, len(f) // 50)
    if mode == 'upper':
        peaks, _ = find_peaks(Pxx, distance=min_dist)
    else:
        peaks, _ = find_peaks(-Pxx, distance=min_dist)
    locs = f[peaks]
    pks = Pxx[peaks]
    if len(locs) < 2:
        return f, np.full_like(f, np.mean(Pxx))
    # 强制包含首尾频率点，避免外推
    locs = np.unique(np.concatenate([[f[0]], locs, [f[-1]]]))
    pks = np.interp(locs, f[peaks], Pxx[peaks])  # 用线性插值获取端点的包络值
    # 改用 Pchip 保持单调不振荡
    pchip = PchipInterpolator(locs, pks, extrapolate=True)
    return f, pchip(f)

f_upper_raw, P_upper_raw = get_envelope(f, Pxx_raw, 'upper')
f_lower_raw, P_lower_raw = get_envelope(f, Pxx_raw, 'lower')
f_upper_filt, P_upper_filt = get_envelope(f, Pxx_filt, 'upper')
f_lower_filt, P_lower_filt = get_envelope(f, Pxx_filt, 'lower')

# ===== 绘图 =====
plt.figure(figsize=(10, 6.5))
ax = plt.gca()
ax.set_yscale('log')

# 降噪后填充
plt.fill_between(f_upper_filt, P_upper_filt, P_lower_filt, color='r', alpha=0.08)
plt.plot(f_upper_filt, P_upper_filt, 'r--', linewidth=1.0, label='降噪后上包络')
plt.plot(f_lower_filt, P_lower_filt, 'r--', linewidth=1.0, label='降噪后下包络')
plt.plot(f, Pxx_filt, 'r-', linewidth=1.2, label='降噪后功率谱')

# 降噪前填充
plt.fill_between(f_upper_raw, P_upper_raw, P_lower_raw, color='b', alpha=0.08)
plt.plot(f_upper_raw, P_upper_raw, 'b--', linewidth=1.0, label='降噪前上包络')
plt.plot(f_lower_raw, P_lower_raw, 'b--', linewidth=1.0, label='降噪前下包络')
plt.plot(f, Pxx_raw, 'b-', linewidth=1.2, label='降噪前功率谱')

plt.xlabel('频率 (周期/序号)')
plt.ylabel('功率谱密度')
plt.title(f'降噪前后傅里叶频谱对比 (信噪比 = {SNR:.2f} dB)')
plt.legend(loc='best')
plt.grid(True, which='both', alpha=0.3)
plt.tight_layout()

save_path = os.path.join(output_dir, 'spectrum_comparison.png')
plt.savefig(save_path, dpi=300)
plt.close()
print(f'图片已保存至 {save_path}')